### PR TITLE
Feature: z-layering of windows

### PR DESCRIPTION
**Pull request by tcld**
After a great example by @senthilbaboo (#2144), I reworked the (second) attempt at layering code (#2029) to make better use of X-functionality. This is the first implementation I finished and tested, so there will be some rough edges and room for improvement. I also have yet to see how the tests fare.

The `core.manager.change_layer()`-method is at the heart of this PR; things seems to work as well as the previous implementation so far (even a little better, since #676 seems to be fixed by this), but so far I don't really like the spaghettis in that method, especially considering that there are more X-flags we should be taking care of.

- [x] convert `_Window.cmd_keep_above()` and `_Window.cmd_keep_below()` to properties (keep the methods for easy access, though)

### CODE REVIEW

File: libqtile/backend/x11/xcbq.py
Comment (m-col) [inline libqtile/backend/x11/xcbq.py:933]: We already have a similar `query_tree` method on the Window class, used by the root window's instance of that class. Would it be feasible to adapt that one for your purpose, or your version here for the other one's purpose? Otherwise there are two apparent ways of doing what appears to be the same thing.

File: libqtile/core/manager.py
Comment (m-col) [inline libqtile/core/manager.py:428]: Could you move the X-specific logic into the core itself, keeping only general high-level logic in the manager?

Also with the stuff below.

File: libqtile/core/manager.py
Comment (m-col) [inline libqtile/core/manager.py:453]: Should this also return if `len(stack)` equals 0?

File: libqtile/core/manager.py
Comment (m-col) [inline libqtile/core/manager.py:427]: You might be able to do `key=stack.index` instead of `key=lambda wid: stack.index(wid)`

**Review by m-col — CHANGES_REQUESTED** [review]
This looks super nice (haven't tested yet). Most of it looks fine, I'd just suggest keeping things in the `Qtile` class high-level and moving as much X-specific logic into the X core as possible.

From your point of view, what stage is this at? By the looks of the commits it looks "done".

File: libqtile/backend/x11/xcbq.py
Comment (tcld) [reply libqtile/backend/x11/xcbq.py:933]: Yeah, I wasn't sure about this. The other method wraps the window ids with newly created windows, which seems useless overhead for the purposes of this PR. If that wasn't the case, one could just call the existing method on the root window and we wouldn't need a second implementation.

The original `query_tree` is only called in one place, I think (haven't looked through the tests, though). So it should be possible to find some common ground.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py:428]: So... just call `query_tree` in the core and directly feed the result into the stacking order? Without anything in between? That would probably be much easier, actually.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py:453]: It should, but since we should have at least one window (which this method is being called from), an exception might be interesting to notice.

**Comment by tcld** [comment]
About what stage this is at: Since I had a pretty solid template both with the previous PR (which worked nicely for months) and the code by @senthilbaboo, this should just need a little bit of polish - thank you very much for the feedback. :)

File: libqtile/core/manager.py
Comment (ramnes) [inline libqtile/core/manager.py:473]: Why not moving that logic into each of the methods rather than handling it in a single method?

File: libqtile/core/manager.py
Comment (ramnes) [inline libqtile/core/manager.py:430]: Why adding this method to the manager rather than the Window object? Especially if its main argument is a window id.

File: libqtile/core/manager.py
Comment (ramnes) [inline libqtile/core/manager.py:480]: We should not play with xcffib outside the X11 backend files. I think this belongs to `xcbq.Window.configure`, feel free to modify it to your needs.

File: libqtile/window.py
Comment (ramnes) [inline libqtile/window.py:472]: We're already doing two ConfigureWindow in two lines, let's not add a third one. :)

File: libqtile/window.py
Comment (ramnes) [inline libqtile/window.py:992]: Why adding the `above` argument to the function rather than just switching to False here?

File: libqtile/window.py
Comment (ramnes) [inline libqtile/window.py:1405]: I'm cherry picking this into master, please rebase.

File: test/test_manager.py
Comment (ramnes) [inline test/test_manager.py:957]: I guess we don't need this anymore :)

File: test/test_manager.py
Comment (ramnes) [inline test/test_manager.py:952]: You can use the `display` fixture instead, now.

File: libqtile/core/manager.py
Comment (ramnes) [inline libqtile/core/manager.py:427]: Why doing this rather than mapping into query_tree directly? Or writing another method for this specifically, that could almost replace _clients() in the tests if I'm understanding it correctly.

**Review by ramnes — CHANGES_REQUESTED** [review]
This is getting much more understandable and clean than the two previous PR, thank you so much for all the work you're putting into this, and thank you @senthilbaboo for putting us back on the right track. :)

Let's do it!

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py:430]: Since the logic shuffles all windows around, I wouldn't really expect it in the `Window`-class. It also requires access to `windows_map`.  I'll think about it some more.

File: libqtile/window.py
Comment (tcld) [reply libqtile/window.py:472]: I don't know what happens in the backend - is this worth thinking about?

In general I agree, but the only solution I see at the moment would be that `change_layer` just helps gather parameters for the eventual call to `configure` instead of actively changing the layer. Do you have something better in mind?

File: libqtile/window.py
Comment (tcld) [reply libqtile/window.py:992]: As far as I remember, the overall reasoning for adding the parameter at all was that without it moving floating windows around got very expensive (they were layered over and over for every pixel they had been moved).
At that point I didn't dare to mess with this too much, since the effects are a bit unclear. I'll give it a go locally.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py:427]: Thank you!

File: CHANGELOG
Comment (tcld) [reply CHANGELOG:67]: I chose `cmd_move_up` and `cmd_move_down` now. I assume `cmd_keep_above` etc.

File: test/test_manager.py
Comment (tcld) [reply test/test_manager.py:952]: Sorry, I just borrowed the setup lines in here from other places in this file. I am having a somewhat hard time with the tests, the typeless-ness of Python makes understanding of some of these lines impossible outside of a debugger (and I have no setup at the moment to debug tests).

So your two comments on the test confuse me. Could you maybe elaborate a bit?

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py]: To keep the functionality as is (I'm not sure it would be a good idea to just feed the full stack back into X's stacking order), I just moved the three middle lines into the called core function with the latest change to the first commit. Let me know if you agree or disagree with it.

File: CHANGELOG
Comment (tcld) [reply CHANGELOG]: Maybe `cmd_lift` and `cmd_lower` might be even better?

File: libqtile/backend/x11/xcbq.py
Comment (tcld) [reply libqtile/backend/x11/xcbq.py:933]: I merged the functions. Next push will contain the changes.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py]: I reworked the function completely, now this comparison is a `< 2`.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py]: I reworked the entire function. If I am not overlooking anything, the current solution is much more elegant.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py:427]: I'm still not sure about this. We need access to potentially all other windows to find our new place in the stack. Is that something I should be accessing from a `Window`? If you think so, I'll move the method.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py:477]: I agree, I'll see if I can modify `configure` to be suitable for both purposes.

File: libqtile/window.py
Comment (tcld) [reply libqtile/window.py]: Done.

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py]: After a comment by @m-col I moved the core of this function into `backend/x11/core.py` while also simplifying it to

```python
_, _, stack = self._root.query_tree()
windows = [w for w in stack if w in windows]
```

I assume that covers your comment, too, since the entire sorting was superfluous and has been removed.

File: CHANGELOG
Comment (ramnes) [reply CHANGELOG]: I like move_down and move_up

Comment (ramnes) [reply CHANGELOG]: Wait, no, this might conflict with the logic behind #1875 some day. What do you think of move_below and move_above? That would be consistent with the two other methods.

**Comment by tcld** [comment]
This is currently entirely broken, since I was making use of the `net_wm_...`-flags a lot, but they have been removed recently. Haven't had the time yet to find out how to set the flags now.

**Comment by m-col** [comment]
Huh, sorry about that. They haven't yet been used in-tree which is why they were removed. Feel free to add a way to access the ones you need for the PR.

Just reading through that code again I realised that (NetWmState)[https://github.com/qtile/qtile/blob/master/libqtile/backend/x11/xcbq.py#L401] is not used elsewhere in libqtile either, so should probably have been removed at the same time. Reading through that class will probably give you a good idea how to access those properties. Indeed all you need to do is do `win.window.get_property('_NET_WM_STATE_X', 'ATOM', unpack=int)` for whichever properties you are interested in.

**Comment by tcld** [comment]
I think I noted that `NetWmState` isn't used anymore in a comment to your PR, did I not?

**Comment by m-col** [comment]
Oh yes, you did!

File: libqtile/core/manager.py
Comment (tcld) [reply libqtile/core/manager.py]: I modified the call and it seems there were no further modifications needed in `xcbq.Window.configure`. So I'm going to mark this resolved.

File: libqtile/window.py
Comment (tcld) [reply libqtile/window.py]: The current implementation avoids the third call. Please have a look and let me know whether you really prefer it - I'm not decided on it yet.^^

**Comment by tcld** [comment]
I still have a few of your comments to consider, but at least the PR is now working again. I have been sorely missing the pinning functionality on my machine.

As far as I see it, the only real difficulty about this entire implementation is inside `core.manager.change_layer`. Climbing up (or down) the window tree while deciding which windows one can move beyond and which not has proven far more complex than I thought. The different EWMH-flags seem managable, but child-windows complicate the situation quite a bit.

Maybe I should just provide the most basic implementation (so the overall behaviour stays as it currently is) and then the method can be fleshed out in another PR.

**Comment by kamiox** [comment]
@tcld I'm getting this crash when using your PR.
I never had this issue on the upstream Qtile but I need to investigate what are the steps to reproduce it.

>  2021-03-16 19:08:50,320 ERROR libqtile core.py:_xpoll():L305 Got an exception in poll loop
> Traceback (most recent call last):
>   File "~/qtile/libqtile/backend/x11/core.py", line 273, in _xpoll
>     ret = target(event)
>   File "~/qtile/libqtile/window.py", line 1202, in handle_PropertyNotify
>     self.update_hints()
>   File "~/qtile/libqtile/window.py", line 282, in update_hints
>     self.group.layout_all()
>   File "~/qtile/libqtile/group.py", line 164, in layout_all
>     self.floating_layout.layout(floating, screen_rect)
>   File "~/qtile/libqtile/layout/base.py", line 57, in layout
>     self.configure(i, screen_rect)
>   File "~/qtile/libqtile/layout/floating.py", line 288, in configure
>     client.place(
>   File "~/qtile/libqtile/window.py", line 469, in place
>     self.qtile.change_layer(self.window.wid, True, kwarg)
>   File "~/qtile/libqtile/core/manager.py", line 467, in change_layer
>     if stack.index(wid) == pos:
> ValueError: 107045453 is not in list

**Comment by tcld** [comment]
Thank you for testing. That looks like you are trying to raise a window that X doesn't know about. Does the crash appear randomly or can you relate it to a specific program you are using?

**Comment by kamiox** [comment]
> Thank you for testing. That looks like you are trying to raise a window that X doesn't know about. Does the crash appear randomly or can you relate it to a specific program you are using?

These crashes have happened when I was using Android Studio. But I don't know a particular way to reproduce it.

File: CHANGELOG
Comment (tcld) [reply CHANGELOG]: Renamed.

File: libqtile/window.py
Comment (tcld) [reply libqtile/window.py:993]: I am setting this to `False` now. Have to test whether `ScratchPad`s still work properly, though. But they seem to have a few other issues at the moment anyway, so spending some time might be worth it.

Comment (tcld) [reply libqtile/window.py]: This seems fine. A pinned window stayed above my `ScratchPad`, in all other cases the `ScratchPad` seemed to win out.

**Comment by tcld** [comment]
I made a few small changes to the layering function while
a) implementing a few of the above suggestions and
b) splitting up the commits a little further to separate the foundation code from the stuff building on it.

@kamiox : If you could find the time to test this again, I would highly appreciate it. I couldn't reproduce your issue, but at least "non existing" windows shouldn't be raised anymore.

**Comment by tcld** [comment]
I finally got around to fixing the tests again. For now I left out moving child-windows around when a parent is moved, I'll leave that for a potential future PR. It might even be better to not handle that situation at all. (The current code does, however, disallow to lower child windows.)

@ramnes: I never really understood what you meant with the comments on the test-code.

**Comment by kamiox** [comment]
>  If you could find the time to test this again, I would highly appreciate it. I couldn't reproduce your issue, but at least "non existing" windows shouldn't be raised anymore.

@tcld  I have tested yesterday's changes and it worked quite well without any crashes and child windows were properly raised above their parents. I haven't tested it extensively but the overall experience was very good, it worked for me much better than the version from the master branch or even when compared to other tiling window managers.

However, after pulling today's changes something go wrong. Most of the child windows (for instance file chooser dialogs) are hidden below their parents.

**Comment by tcld** [comment]
Thanks for the feedback. I had only tested the layering with a few applications and found no issues. After I read your message I tried the file chooser dialog in VSCode, and it is indeed opened behind the parent window (when opened for the first time). I'll have another look.

**Comment by tcld** [comment]
Issue found and fixed.

**Comment by tcld** [comment]
I (re-)added the ability to display multiple windows on top of each other in the `Max`-layout.

The code can probably be made prettier, but the functionality has proven very solid over the past few days.

**Comment by tcld** [comment]
Here are the parts of my config that handle the PIP-functionality, in case somebody wants to try that out. This is particularly useful for video windows that run in "full screen" (i.e. within the fake screen):

```python
fake_screens = [  # we have to use fake_screens instead of screens
  Screen(x=0, y=0, width=1920, height=1080), # actual screen
  Screen(x=1280, y=720, width=640, height=360)  # overlayed fake screen (bottom right)
]

def shift_to_pip(group):
    def f(qtile):
        _from = qtile.current_group
        _to = qtile.groups_map[group]
        if _from == _to:
            return
        client = _from.current_window
        for c in list(_to.windows):  # empty out target group
            c.kept_above = False
            c.togroup(_from.name, switch_group=False)

        if not client:
            return
        client.togroup(group, switch_group=False)
        client.kept_above = True

    return f

def remove_all_pip(group):
    def f(qtile):
        _to = qtile.current_group
        _from = qtile.groups_map[group]
        if _from == _to:
            return
        for c in list(_from.windows):
            c.kept_above = False
            c.togroup(_to.name, switch_group=False)

    return f

groups.extend([
    Group('pip', position=10, label='PIP', layout='max')
])

keys.extend([
    Key([mod], 'p', lazy.function(shift_to_pip('pip')), desc='Move focused window to group PIP.'),
    Key([mod, alt], 'p', lazy.function(remove_all_pip('pip')), desc='Move all PIP-windows to current group.'),
])

# rough example of a GroupBox-widget that excludes the PIP-group
widget.GroupBox(visible_groups=[g.name for g in groups if g.name not in ['pip']])
```

**Comment by m-col** [comment]
Could you rebase? `libqtile/window.py` has moved now into the x11 backend but I would imagine that many of the changes don't need much editing to work again.

Don't worry about implementing the Wayland side of Z stacking if you don't want to; I'm not completely sure how it would look, but it would probably make the most sense and be the easiest approach to get it working in X and then iterate on top of that implementation to get it working for Wayland. Indeed proper Z layering is very important for the Wayland backend's usability (I mean, it is for x11, but more so for wayland because at least in X the X server does some of it), so I'm eager to put more focus on this if you're up for it!

**Comment by tcld** [comment]
I'll have to take some time looking at the conflicts, they don't appear to be trivial this time.

**Comment by tcld** [comment]
A lot of things have been moved, so it took me almost two hours just to rebase the first commit. Maybe dropping this is the smarter approach; I'll just keep running my local version for the time being.

**Comment by m-col** [comment]
OK, no worries. Apologies! If you'd like I can take a shot at sticking your changes on top of the current head and re-submit?

**Comment by tcld** [comment]
Nothing to apologize for, I am very happy to see us moving towards Wayland. :)

And yes, please have a go! You likely know where to put the code, so you can probably do it in a reasonable timeframe. Although I think there will be some additional work needed to separate the `change_layer`-method to really keep all the X11-related stuff out of `core/manager.py`. This wasn't much of an issue for the first commit, but it will likely get more complicated for the later ones.

**Comment by m-col** [comment]
What was the rationale for commit `Move keep_{above,below}-code into properties `?

**Comment by tcld** [comment]
It adds the ability to read and explicitly set those states, which the simple functions `keep_above()` etc. won't do.

**Comment by m-col** [comment]
I was playing with pushing to a fork but looks like I don't have write permission for your branch. I have rebased this PR on top current master and I'm pretty sure everything survived unscathed. No problem at all if you aren't, @tcld, but would be interested in testing out the rebased version and seeing how it is? I can submit it as a separate PR or push to this PR if you give me access. I'm also happy to open as a separate PR and work on this independently.

**Comment by tcld** [comment]
I actually don't know how to give you access to my branch. But feel free to create your own branch, it's likely all your work anyway. :)

I would be happy to give your changes a test run, although I'd have to update Qtile first (couldn't live without the layering, so I haven't updated for several months now). And I'd need at least a few days for it, since I'm going to be busy at least all of tomorrow.

**Comment by m-col** [comment]
> I actually don't know how to give you access to my branch. But feel free to create your own branch, it's likely all your work anyway. :)

I actually didn't need to make many changes, just move things around! I've never done the push-to-PR thing before so kinda want to try it out. I found this at bottom of the side bar on a PR I made on another repository (where I'm not a maintainer). Maybe you see the same on this page, but unchecked?

image

> I would be happy to give your changes a test run, although I'd have to update Qtile first (couldn't live without the layering, so I haven't updated for several months now). And I'd need at least a few days for it, since I'm going to be busy at least all of tomorrow.

Brilliant! I'm very eager to get this working as I don't want to put you out again, but there is no rush at all so that's all good.

**Comment by tcld** [comment]
The description says that you are allowed to add commits - although I guess you want to do a forced push, right? Maybe that's not possible.

**Comment by ramnes** [comment]
You can force push it it's checked. @m-col, while it's very possible to do this by hand, I recommend using GitHub's CLI for this. Just run `gh co 2194`, throw your modifications and push, the tool will do all the heavy work for you.

**Comment by m-col** [comment]
Cool, got it, thanks.

**Comment by tcld** [comment]
Raising and lowering windows doesn't seem to work at the moment:

```log
2021-09-05 14:38:18,414 ERROR libqtile manager.py:process_key_event():L361 KB command error keep_above: Traceback (most recent call last):
  File "/home/tcl/Software/qtile/libqtile/command/interface.py", line 313, in call
    return SUCCESS, cmd(*args, **kwargs)
  File "/home/tcl/Software/qtile/libqtile/backend/x11/window.py", line 1944, in cmd_keep_above
    self.kept_above = not self.kept_above
  File "/home/tcl/Software/qtile/libqtile/backend/x11/window.py", line 1456, in kept_above
    atom = self.qtile.conn.atoms['_NET_WM_STATE_ABOVE']
AttributeError: 'Qtile' object has no attribute 'conn'
```

Not sure what has changed about the `Qtile`-object.

**Comment by m-col** [comment]
Ah, the `conn` was removed as it's X11-specific, but it can be accessed at `self.qtile.core.conn`, so just that `core` bit needs to be added.

**Comment by tcld** [comment]
With that I stop getting an exception, but nothing is raised or lowered anyway.

**Comment by m-col** [comment]
I fixed that conn reference and rebased to fix that conflict. Also getting no working behaviour :(

**Comment by m-col** [comment]
It's because the line in `change_layer` does `query_tree()` on the window rather than root, so that bit should be `self.qtile.core._root.query_tree()` rather than `self.window.query_tree()`.

**Comment by tcld** [comment]
Yes, that seems to have fixed it. :)

**Comment by tcld** [comment]
Switching layers (in `Max`-layout) and pinning windows to the bottom of the stack doesn't work. The latter one does _something_, but I can't really tell what it is.

**Comment by m-col** [comment]
It seems the `kept_above` but not the `kept_below` property call `self.change_layer()`. Should the latter call it with `self.change_layer(up=False)`? How should `change_layer` be interpreted - e.g. is it "update position in response to a change in layering?", or telling the window to change layer (which would then be similar to `move_{above,below}`)?

**Comment by tcld** [comment]
It's more the latter (there is a bit of documentation belonging to `change_layer`). So if you say `change_layer(up=true)` you are asking the WM to move you to the top with respect to windows that live in the same layer as you. So you will become the one-eyed king of the one-eyed people, but you won't touch the blind or the two-eyed ones. `change_layer(up=false)` does the opposite.

The doc string mentions a number of edge cases.

**Comment by m-col** [comment]
When a windows WM_STATE changes to specify that the window should be kept above other windows with the hint, we are then calling `change_layer`, but should we not instead be calling a method that re-sorts the windows according to their states?

**Comment by tcld** [comment]
I guess it does both. You can modify flags or not, and when you call `change_layer`, everything is going to be put into place properly.

**Comment by m-col** [comment]
Huh, this PR says both "changes requested" but also that it is waiting for review. Which is it, github?

I'm going to assume that it is the latter and that I should give this another review and test. I'll make this top priority and give it a look over in the next couple of days @tcld! Any additional testing would be really beneficial as we should really get this merged, so @elParaguayo if you'd be happy to have this in your day-to-day setup as you're running X then that would be great!

**Comment by elParaguayo** [comment]
OK - I'll take a look.

**Comment by elParaguayo** [comment]
Yeah. None of those commands are doing anything for me.

**Comment by elParaguayo** [comment]
I think this might have broken my scratchpads. They won't open on top of my browser and, instead, open behind it.

**Comment by elParaguayo** [comment]
So I've actually had this PR running on my setup for a while now and I haven't seen any day-to-day issues which is great.

I don't use floating windows much but I have played around a bit with the various commands and have a few minor observations:
`cmd_move_above` - this moves to the top rather than moving above the next window in the stack. I need to look at the logic for this a bit more closely. Same issue for `cmd_move_below`.
`kept_above` currently overrides `move_below` so moving has no impact. That's a design choice rather than a bug but I was just wondering if users might expect the window to move if they do call the `move_below` command (i.e. we deactivate `kept_above`).

I'm happy to make changes. Just looking for some input.

**Comment by elParaguayo** [comment]
This PR fixes #1145.

**Comment by m-col** [comment]
> Also, should floating windows default to being above tiled windows?

If you mean having something in the default config then yeah - like mentioned in https://github.com/qtile/qtile/issues/3120#issuecomment-1019288990

Even if this just exists as a hook defined in the default confit then that would be useful.

> kept_above currently overrides move_below so moving has no impact. That's a design choice rather than a bug but I was just wondering if users might expect the window to move if they do call the move_below command (i.e. we deactivate kept_above).

Hmm, perhaps we can use a `force=True` optional argument to do the override? I can see how both behaviours might be desirable.

**Comment by elParaguayo** [comment]
Closing this in favour of #3409

**Pull request closed without merging**
